File: js/src/client.test.ts
import { describe, it, expect, beforeEach, vi } from 'vitest';
import { FestivoClient } from './client';

describe('FestivoClient', () => {
  let client: FestivoClient;

  beforeEach(() => {
    client = new FestivoClient({ apiKey: 'test-key' });
  });

  describe('initialization', () => {
    it('should initialize with default base URL', () => {
      expect(client.baseUrl).toBe('https://api.getfestivo.com');
      expect(client.apiKey).toBe('test-key');
    });

    it('should accept custom base URL', () => {
      const customClient = new FestivoClient({
        apiKey: 'test-key',
        baseUrl: 'https://custom.api.com',
      });
      expect(customClient.baseUrl).toBe('https://custom.api.com');
    });

    it('should work without API key', () => {
      const noKeyClient = new FestivoClient();
      expect(noKeyClient.apiKey).toBeUndefined();
    });
  });

  describe('getHolidays', () => {
    it('should call correct endpoint with parameters', async () => {
      global.fetch = vi.fn().mockResolvedValue({
        ok: true,
        json: async () => ({ holidays: [], total: 0 }),
      });

      await client.getHolidays('US', 2026);

      expect(fetch).toHaveBeenCalledWith(
        expect.stringContaining('/v3/public-holidays/list'),
        expect.objectContaining({
          headers: expect.objectContaining({
            Accept: 'application/json',
            Authorization: 'Bearer test-key',
          }),
        })
      );

      const callUrl = (fetch as any).mock.calls[0][0];
      expect(callUrl).toContain('country=US');
      expect(callUrl).toContain('year=2026');
    });

    it('should include optional parameters', async () => {
      global.fetch = vi.fn().mockResolvedValue({
        ok: true,
        json: async () => ({ holidays: [], total: 0 }),
      });

      await client.getHolidays('GB', 2026, {
        regions: 'GB-SCT',
        type: 'public',
        language: 'en',
      });

      const callUrl = (fetch as any).mock.calls[0][0];
      expect(callUrl).toContain('regions=GB-SCT');
      expect(callUrl).toContain('type=public');
      expect(callUrl).toContain('language=en');
    });

    it('should throw on API error', async () => {
      global.fetch = vi.fn().mockResolvedValue({
        ok: false,
        status: 401,
        statusText: 'Unauthorized',
      });

      await expect(client.getHolidays('US', 2026)).rejects.toThrow(
        'API error: 401 Unauthorized'
      );
    });
  });

  describe('getCityHolidays', () => {
    it('should call getHolidays with city code in regions', async () => {
      global.fetch = vi.fn().mockResolvedValue({
        ok: true,
        json: async () => ({ holidays: [], total: 0 }),
      });

      await client.getCityHolidays('IT', 'IT-MILAN', 2026);

      const callUrl = (fetch as any).mock.calls[0][0];
      expect(callUrl).toContain('country=IT');
      expect(callUrl).toContain('regions=IT-MILAN');
      expect(callUrl).toContain('year=2026');
    });

    it('should support additional options', async () => {
      global.fetch = vi.fn().mockResolvedValue({
        ok: true,
        json: async () => ({ holidays: [], total: 0 }),
      });

      await client.getCityHolidays('IT', 'IT-ROME', 2026, {
        type: 'public',
        language: 'it',
      });

      const callUrl = (fetch as any).mock.calls[0][0];
      expect(callUrl).toContain('type=public');
      expect(callUrl).toContain('language=it');
    });
  });

  describe('getRegionalHolidays', () => {
    it('should call getHolidays with region code', async () => {
      global.fetch = vi.fn().mockResolvedValue({
        ok: true,
        json: async () => ({ holidays: [], total: 0 }),
      });

      await client.getRegionalHolidays('GB', 'GB-SCT', 2026);

      const callUrl = (fetch as any).mock.calls[0][0];
      expect(callUrl).toContain('country=GB');
      expect(callUrl).toContain('regions=GB-SCT');
      expect(callUrl).toContain('year=2026');
    });
  });

  describe('checkHoliday', () => {
    it('should call check endpoint with correct parameters', async () => {
      global.fetch = vi.fn().mockResolvedValue({
        ok: true,
        json: async () => ({ is_holiday: true }),
      });

      await client.checkHoliday('US', '2026-12-25');

      const callUrl = (fetch as any).mock.calls[0][0];
      expect(callUrl).toContain('/v3/public-holidays/list/check');
      expect(callUrl).toContain('country=US');
      expect(callUrl).toContain('date=2026-12-25');
    });

    it('should support optional regions parameter', async () => {
      global.fetch = vi.fn().mockResolvedValue({
        ok: true,
        json: async () => ({ is_holiday: true }),
      });

      await client.checkHoliday('IT', '2026-12-07', { regions: 'IT-MILAN' });

      const callUrl = (fetch as any).mock.calls[0][0];
      expect(callUrl).toContain('regions=IT-MILAN');
    });
  });

  describe('type safety', () => {
    it('should return properly typed HolidaysResponse', async () => {
      const mockResponse = {
        holidays: [
          {
            date: '2026-01-01',
            name: 'New Year',
            type: 'public',
            observed: '2026-01-01',
            public: true,
            country: 'US',
            subdivisions: [],
          },
        ],
        total: 1,
      };

      global.fetch = vi.fn().mockResolvedValue({
        ok: true,
        json: async () => mockResponse,
      });

      const result = await client.getHolidays('US', 2026);
      expect(result).toEqual(mockResponse);
      expect(Array.isArray(result.holidays)).toBe(true);
      expect(typeof result.total).toBe('number');
    });

    it('should return properly typed CheckHolidayResponse', async () => {
      const mockResponse = {
        is_holiday: true,
        holiday: {
          date: '2026-12-25',
          name: 'Christmas',
          type: 'public',
          observed: '2026-12-25',
          public: true,
          country: 'US',
          subdivisions: [],
        },
      };

      global.fetch = vi.fn().mockResolvedValue({
        ok: true,
        json: async () => mockResponse,
      });

      const result = await client.checkHoliday('US', '2026-12-25');
      expect(result).toEqual(mockResponse);
      expect(typeof result.is_holiday).toBe('boolean');
    });
  });
});

File: go/festivo/client_test.go
package festivo

import (
	"encoding/json"
	"net/http"
	"net/http/httptest"
	"os"
	"testing"
)

func TestNewClient(t *testing.T) {
	client := NewClient("test-key")

	if client.BaseURL != "https://api.getfestivo.com" {
		t.Errorf("Expected base URL to be https://api.getfestivo.com, got %s", client.BaseURL)
	}

	if client.APIKey != "test-key" {
		t.Errorf("Expected API key to be test-key, got %s", client.APIKey)
	}

	if client.HTTP == nil {
		t.Error("Expected HTTP client to be initialized")
	}
}

func TestGetHolidays(t *testing.T) {
	// Create mock server
	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		// Verify request
		if r.Method != "GET" {
			t.Errorf("Expected GET request, got %s", r.Method)
		}

		if r.Header.Get("Authorization") != "Bearer test-key" {
			t.Errorf("Expected Authorization header with bearer token")
		}

		if r.Header.Get("Accept") != "application/json" {
			t.Errorf("Expected Accept header to be application/json")
		}

		// Check query parameters
		query := r.URL.Query()
		if query.Get("country") != "US" {
			t.Errorf("Expected country to be US, got %s", query.Get("country"))
		}
		if query.Get("year") != "2026" {
			t.Errorf("Expected year to be 2026, got %s", query.Get("year"))
		}

		// Return mock response
		resp := HolidaysResponse{
			Holidays: []Holiday{
				{
					Date:     "2026-01-01",
					Name:     "New Year's Day",
					Type:     "public",
					Observed: "2026-01-01",
					Public:   true,
					Country:  "US",
				},
			},
			Total: 1,
		}
		json.NewEncoder(w).Encode(resp)
	}))
	defer server.Close()

	// Create client with mock server
	client := NewClient("test-key")
	client.BaseURL = server.URL

	// Test
	result, err := client.GetHolidays("US", 2026, nil)
	if err != nil {
		t.Fatalf("Expected no error, got %v", err)
	}

	if result.Total != 1 {
		t.Errorf("Expected total to be 1, got %d", result.Total)
	}

	if len(result.Holidays) != 1 {
		t.Fatalf("Expected 1 holiday, got %d", len(result.Holidays))
	}

	if result.Holidays[0].Name != "New Year's Day" {
		t.Errorf("Expected holiday name to be New Year's Day, got %s", result.Holidays[0].Name)
	}
}

func TestGetHolidaysWithOptions(t *testing.T) {
	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		query := r.URL.Query()
		
		if query.Get("regions") != "GB-SCT" {
			t.Errorf("Expected regions to be GB-SCT, got %s", query.Get("regions"))
		}
		if query.Get("type") != "public" {
			t.Errorf("Expected type to be public, got %s", query.Get("type"))
		}
		if query.Get("language") != "en" {
			t.Errorf("Expected language to be en, got %s", query.Get("language"))
		}

		resp := HolidaysResponse{Holidays: []Holiday{}, Total: 0}
		json.NewEncoder(w).Encode(resp)
	}))
	defer server.Close()

	client := NewClient("test-key")
	client.BaseURL = server.URL

	opts := &HolidayOptions{
		Regions:  "GB-SCT",
		Type:     "public",
		Language: "en",
	}

	_, err := client.GetHolidays("GB", 2026, opts)
	if err != nil {
		t.Fatalf("Expected no error, got %v", err)
	}
}

func TestGetCityHolidays(t *testing.T) {
	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		query := r.URL.Query()
		
		if query.Get("country") != "IT" {
			t.Errorf("Expected country to be IT, got %s", query.Get("country"))
		}
		if query.Get("regions") != "IT-MILAN" {
			t.Errorf("Expected regions to be IT-MILAN, got %s", query.Get("regions"))
		}
		if query.Get("year") != "2026" {
			t.Errorf("Expected year to be 2026, got %s", query.Get("year"))
		}

		resp := HolidaysResponse{Holidays: []Holiday{}, Total: 0}
		json.NewEncoder(w).Encode(resp)
	}))
	defer server.Close()

	client := NewClient("test-key")
	client.BaseURL = server.URL

	_, err := client.GetCityHolidays("IT", "IT-MILAN", 2026, nil)
	if err != nil {
		t.Fatalf("Expected no error, got %v", err)
	}
}

func TestGetRegionalHolidays(t *testing.T) {
	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		query := r.URL.Query()
		
		if query.Get("country") != "GB" {
			t.Errorf("Expected country to be GB, got %s", query.Get("country"))
		}
		if query.Get("regions") != "GB-SCT" {
			t.Errorf("Expected regions to be GB-SCT, got %s", query.Get("regions"))
		}

		resp := HolidaysResponse{Holidays: []Holiday{}, Total: 0}
		json.NewEncoder(w).Encode(resp)
	}))
	defer server.Close()

	client := NewClient("test-key")
	client.BaseURL = server.URL

	_, err := client.GetRegionalHolidays("GB", "GB-SCT", 2026, nil)
	if err != nil {
		t.Fatalf("Expected no error, got %v", err)
	}
}

func TestCheckHoliday(t *testing.T) {
	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		query := r.URL.Query()
		
		if query.Get("country") != "US" {
			t.Errorf("Expected country to be US, got %s", query.Get("country"))
		}
		if query.Get("date") != "2026-12-25" {
			t.Errorf("Expected date to be 2026-12-25, got %s", query.Get("date"))
		}

		if r.URL.Path != "/v3/public-holidays/check" {
			t.Errorf("Expected path to be /v3/public-holidays/check, got %s", r.URL.Path)
		}

		resp := CheckHolidayResponse{
			IsHoliday: true,
			Holiday: &Holiday{
				Date:     "2026-12-25",
				Name:     "Christmas Day",
				Type:     "public",
				Observed: "2026-12-25",
				Public:   true,
				Country:  "US",
			},
		}
		json.NewEncoder(w).Encode(resp)
	}))
	defer server.Close()

	client := NewClient("test-key")
	client.BaseURL = server.URL

	result, err := client.CheckHoliday("US", "2026-12-25", "")
	if err != nil {
		t.Fatalf("Expected no error, got %v", err)
	}

	if !result.IsHoliday {
		t.Error("Expected is_holiday to be true")
	}

	if result.Holiday.Name != "Christmas Day" {
		t.Errorf("Expected holiday name to be Christmas Day, got %s", result.Holiday.Name)
	}
}

func TestCheckHolidayWithRegions(t *testing.T) {
	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		query := r.URL.Query()
		
		if query.Get("regions") != "IT-MILAN" {
			t.Errorf("Expected regions to be IT-MILAN, got %s", query.Get("regions"))
		}

		resp := CheckHolidayResponse{IsHoliday: true}
		json.NewEncoder(w).Encode(resp)
	}))
	defer server.Close()

	client := NewClient("test-key")
	client.BaseURL = server.URL

	_, err := client.CheckHoliday("IT", "2026-12-07", "IT-MILAN")
	if err != nil {
		t.Fatalf("Expected no error, got %v", err)
	}
}

func TestAPIError(t *testing.T) {
	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		w.WriteHeader(http.StatusUnauthorized)
		w.Write([]byte(`{"error": "Unauthorized"}`))
	}))
	defer server.Close()

	client := NewClient("test-key")
	client.BaseURL = server.URL

	_, err := client.GetHolidays("US", 2026, nil)
	if err == nil {
		t.Error("Expected error for 401 response, got nil")
	}
}

func TestClientWithoutAPIKey(t *testing.T) {
	os.Unsetenv("FESTIVO_API_KEY")
	client := NewClient("")

	if client.APIKey != "" {
		t.Errorf("Expected empty API key, got %s", client.APIKey)
	}
}
